### PR TITLE
[Kotlin cleanup] StudyOptionsFragment

### PR DESCRIPTION
**Pull request by dorrin-sot**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 1]
```diff
@@ -142,21 +142,18 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
...
-        val studyOptionsView = inflater.inflate(R.layout.studyoptions_fragment, container, false)
-        mStudyOptionsView = studyOptionsView
-        mFragmented = requireActivity().javaClass != StudyOptionsActivity::class.java
-        initAllContentViews(studyOptionsView)
-        mToolbar = studyOptionsView.findViewById(R.id.studyOptionsToolbar)
-        if (mToolbar != null) {
-            mToolbar!!.inflateMenu(R.menu.study_options_fragment)
-            configureToolbar()
+        // If container is null, Currently in a layout without a container, so no reason to create our view.
+        return container?.let {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:145]: I prefer the guard clause, it removes context and indentation

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 2]
```diff
@@ -324,46 +325,35 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
...
-                menu.findItem(R.id.action_rebuild).isVisible = true
-                menu.findItem(R.id.action_empty).isVisible = true
-                menu.findItem(R.id.action_custom_study).isVisible = false
-                menu.findItem(R.id.action_deck_or_study_options).setTitle(R.string.menu__study_options)
-            } else {
-                menu.findItem(R.id.action_rebuild).isVisible = false
-                menu.findItem(R.id.action_empty).isVisible = false
-                menu.findItem(R.id.action_custom_study).isVisible = true
-                menu.findItem(R.id.action_deck_or_study_options).setTitle(R.string.menu__deck_options)
+            col != null && col!!.decks.isDyn(col!!.decks.selected()).also { filtered ->
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:327]: Could you extract this to a variable: `filtered`, I'm finding this hard to read (which might be my unfamiliarity with Kotlin)

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 3]
```diff
@@ -324,46 +325,35 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
...
+                menu.findItem(R.id.action_deck_or_study_options).setTitle(
+                    if (filtered) R.string.menu__study_options
+                    else R.string.menu__deck_options
+                )
             }
             // Don't show custom study icon if congrats shown
-            if (mCurrentContentView == CONTENT_CONGRATS) {
-                menu.findItem(R.id.action_custom_study).isVisible = false
-            }
+            menu.findItem(R.id.action_custom_study).isVisible = mCurrentContentView != CONTENT_CONGRATS
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:337]: Awesome

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 4]
```diff
@@ -324,46 +325,35 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
...
+                }
             }
             // Set the back button listener
             if (!mFragmented) {
-                val icon = AppCompatResources.getDrawable(requireContext(), R.drawable.ic_arrow_back_white)
-                icon!!.isAutoMirrored = true
-                mToolbar!!.navigationIcon = icon
+                mToolbar!!.navigationIcon = AppCompatResources.getDrawable(requireContext(), R.drawable.ic_arrow_back_white)!!.apply {
+                    isAutoMirrored = true
+                }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:355]: Can this be made to include the next line?

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 5]
```diff
@@ -421,11 +421,9 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
     }
 
     private fun dismissProgressDialog() {
-        if (mStudyOptionsView != null && mStudyOptionsView!!.findViewById<View?>(R.id.progress_bar) != null) {
-            mStudyOptionsView!!.findViewById<View>(R.id.progress_bar).visibility = View.GONE
-        }
+        mStudyOptionsView?.findViewById<View>(R.id.progress_bar)?.visibility = View.GONE
         // for rebuilding cram decks
-        if (mProgressDialog != null && mProgressDialog!!.isShowing) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:440]: can we add a cleanup to extract this to `.ifShowing { }` or similar. It's a common line of code, and it's messy

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 6]
```diff
@@ -685,11 +685,9 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
          */
         @JvmStatic
         fun newInstance(withDeckOptions: Boolean): StudyOptionsFragment {
-            val f = StudyOptionsFragment()
-            val args = Bundle()
-            args.putBoolean("withDeckOptions", withDeckOptions)
-            f.arguments = args
-            return f
+            return StudyOptionsFragment().apply {
+                arguments = Bundle().apply { putBoolean("withDeckOptions", withDeckOptions) }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:684]: Question: could this be 

`bundleOf("withDeckOptions" to withDeckOptions)`

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt [hunk 7]
```diff
@@ -515,22 +515,11 @@ class StudyOptionsFragment : Fragment(), Toolbar.OnMenuItemClickListener {
...
-                    if (name.isNotEmpty()) {
-                        nameBuilder.append(name[0])
-                    }
-                    if (name.size > 1) {
-                        nameBuilder.append("\n").append(name[1])
-                    }
+                    var name = Decks.path(deck.getString("name"))
                     if (name.size > 3) {
-                        nameBuilder.append("...")
+                        name = arrayOf(name[0], name[1], "...", name.last())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:520]: Nice! Can you add a test here if it's not tested already

**Review by david-allison — COMMENTED** [review]
Thanks!

Can I ask: what's your familiarity with Kotlin? I'm a relative beginner, so there's a few things that might be learning experiences for me. 

Feel free to push back

**Comment by dorrin-sot** [comment]
> Can I ask: what's your familiarity with Kotlin? I'm a relative beginner, so there's a few things that might be learning experiences for me.

I'm actually a beginner too. I started in April while working on this project :)

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt (hunk 5 above)
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:440]: do you mean like this?
```diff
Index: AnkiDroid/src/main/java/com/ichi2/utils/DialogUtils.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/utils/DialogUtils.kt b/AnkiDroid/src/main/java/com/ichi2/utils/DialogUtils.kt
--- a/AnkiDroid/src/main/java/com/ichi2/utils/DialogUtils.kt	(revision 414014ce2c9f90456cf5e66e54a63e44ffd07487)
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/DialogUtils.kt	(date 1655221729425)
@@ -27,4 +27,10 @@
         target.showsDialog = false
         return MaterialDialog.Builder(target.requireContext()).build()
     }
+
+    fun runIfDialogShowing(dialog: MaterialDialog?, runnable: Runnable) {
+        if (dialog?.isShowing == true) {
+            runnable.run()
+        }
+    }
 }
Index: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt	(revision 414014ce2c9f90456cf5e66e54a63e44ffd07487)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt	(date 1655386557731)
@@ -49,6 +49,7 @@
 import com.ichi2.libanki.Decks
 import com.ichi2.libanki.Utils
 import com.ichi2.themes.StyledProgressDialog.Companion.show
+import com.ichi2.utils.DialogUtils
 import com.ichi2.utils.FragmentFactoryUtils.instantiate
 import com.ichi2.utils.HtmlUtils.convertNewlinesToHtml
 import timber.log.Timber
@@ -423,7 +424,7 @@
     private fun dismissProgressDialog() {
         mStudyOptionsView?.findViewById<View>(R.id.progress_bar)?.visibility = View.GONE
         // for rebuilding cram decks
-        if (mProgressDialog?.isShowing == true) {
+        DialogUtils.runIfDialogShowing(mProgressDialog) {
             try {
                 mProgressDialog!!.dismiss()
             } catch (e: Exception) {

```

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:440]: oh is there a way in kotlin to add methods to existing classes?

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt (hunk 1 above)
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:147]: Sorry I didn't understand what do you mean

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt]: Is 65d25c4 ok?

**Comment by mikehardy** [comment]
Oh no - this sat while I was traveling (perhaps you have been busy as well?) and took a bunch of conflicts :weary:

**Pull request closed without merging**